src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,436 @@
+        Vector3d n = cross(Vector3d(0.0, 0.0, 1.0), h);
+
+        double semiMajorAxis = 1.0 / (2.0 / radius.length() - velocity.dot() / GM);
+
+        // Latitude of ascending node
+        double LAN = acos(std::clamp(n.x() / n.length(), -1.0, 1.0));
+        if (n.y() < 0)
+        {
+            LAN = 2.0 * M_PI - LAN;
+        }
+        if (n.length() <= KINDA_SMALL_NUMBER)
+        {
+            LAN = 0.0;
+        }
+
+        const double inclination = acos(h.z() / h.length());
+
+        // Compute true anomaly v and argument of periapsis w
+        double v, w;
+        if (e > KINDA_SMALL_NUMBER)
+        {
+            // True anomaly
+            double m = dot(eVec, radius) / (e * radius.length());
+            v = std::acos(std::clamp(m, -1.0, 1.0));
+            if (dot(radius, velocity) < 0)
+            {
+                v = 2.0 * M_PI - v;
+            }
+            if (m >= 1.0)
+            {
+                v = 0.0;
+            }
+
+            // Argument of periapsis (angle between ascending node and periapsis)
+            w = acos(std::clamp(dot(n, eVec) / (e * n.length()), -1.0, 1.0));
+            if (eVec.z() < 0)
+            {
+                w = 2.0 * M_PI - w;
+            }
+            if (e <= KINDA_SMALL_NUMBER)
+            {
+                w = 0.0;
+            }
+        }
+        else
+        {
+            // If e == 0, the orbit is circular and has no periapsis, so we use argument of latitude as a substitute for argument of periapsis
+            // see https://en.wikipedia.org/wiki/True_anomaly#From_state_vectors
+            double m = dot(n, radius) / (n.length() * radius.length());
+            v = std::acos(std::clamp(m, -1.0, 1.0));
+            if (radius.z() < 0)
+            {
+                v = 2.0 * M_PI - v;
+            }
+            if (m >= 1.0 - KINDA_SMALL_NUMBER)
+            {
+                v = 0.0;
+            }
+            w = 0.0;
+        }
+
+        // Compute eccentric anomaly from true anomaly
+        // use that to find mean anomaly
+        double M0, E;
+        if (e > 1.0 + KINDA_SMALL_NUMBER)
+        {
+            const double F = true_anomaly_to_hyperbolic_eccentric_anomaly(e, v);
+            M0 = e * sinh(F) - F;
+            E = F;
+        }
+        else if (e >= 1.0 - KINDA_SMALL_NUMBER)
+        {
+            // parabolic orbit, todo
+            M0 = 0.5 * std::tan(v / 2.0) + (1 / 6.0) * std::pow(std::tan(v / 2.0), 3);
+            E = v;
+        }
+        else if (e > KINDA_SMALL_NUMBER)
+        {
+            E = true_anomaly_to_eccentric_anomaly(e, v);
+            M0 = E - e * std::sin(E);
+        }
+        else
+        {
+            // circular orbit, todo
+            E = v;
+            M0 = E - e * std::sin(E);
+        }
+
+        return KeplerOrbit(semiMajorAxis, e, inclination, w, LAN, M0, epoch, GM, radius.length(), velocity.length(), E, h.length());
+    }
+
+} // namespace osp
```

Comment:
prefer having a single blank line at the end of files.

Tools like github code review get grumpy without it.